server/src/internalClusterTest/java/org/elasticsearch/indices/IndicesLifecycleListenerIT.java

```diff
@@ -143,12 +143,14 @@ public void beforeIndexCreated(Index index, Settings indexSettings) {
         updateIndexSettings(Settings.builder().put(INDEX_ROUTING_EXCLUDE_GROUP_PREFIX + "._name", node1), "index1");
         ensureGreen("index1");
 
-        var state = clusterAdmin().prepareState().get().getState();
-        logger.info("Final routing is {}", state.getRoutingNodes().toString());
-        var shard = state.routingTable().index("index1").shard(0).primaryShard();
-        assertThat(shard, notNullValue());
-        assertThat(shard.state(), equalTo(ShardRoutingState.STARTED));
-        assertThat(state.nodes().get(shard.currentNodeId()).getName(), equalTo(node1));
+        assertBusy(() -> {
+            var state = clusterAdmin().prepareState().get().getState();
+            var shard = state.routingTable().index("index1").shard(0).primaryShard();
+            assertThat(shard, notNullValue());
+            assertThat(shard.state(), equalTo(ShardRoutingState.STARTED));
+            assertThat(state.nodes().get(shard.currentNodeId()).getName(), equalTo(node1));
+            assertThat(shard.relocationFailureInfo().failedRelocations(), equalTo(5));// see SETTING_ALLOCATION_MAX_RETRY
```

Comment:
why hardcoded and not actual max retry value?

Reply:
I looked at this test before and still trying to understand how these assertions prove that we dont retry forever as test name indicates. Since ensureGreen not reliable as you mentioned in description, reaching maxRetries and still being STARTED on current node not necessarily means we will not retry in the future. I guess this assumption holds until we dont reset retries and MaxRetryAllocationDecider still there, but it's implicit in this test.

Reply:
++ let's say `SETTING_ALLOCATION_MAX_RETRY.get(Settings.EMPTY)` rather than the literal 5.

Also can we `assertBusy()` that the relocation counter reaches this value (or, even better use `org.elasticsearch.test.ClusterServiceUtils#addTemporaryStateListener` to await that cluster state) and then assert that the shard remains `STARTED`?

Reply:
Will update to rely on default rather then a constant.

>that the relocation counter reaches this value

Do you mean the relocation failure counter (one already asserted on L152) or something else?

>and then assert that the shard remains STARTED?

I guess this part is tricky. I can add another assert in the end but theoretically there can always be a scheduled task that issue another cluster state change right after we perform the last assertion. This can not even be excluded by waiting for languid tasks as long as we are not using deterministic queue here.

Reply:
> Do you mean the relocation failure counter (one already asserted on L152)

Yeah, but wait _only_ for that to reach 5, not for the rest of the conditions in the `assertBusy()`.

>  I can add another assert in the end but theoretically there can always be a scheduled task that issue another cluster state change right after we perform the last assertion.

Sure but it shouldn't be relocating that shard once it's reached 5 failures should it?

Reply:
>Sure but it shouldn't be relocating that shard once it's reached 5 failures should it?

That is correct assuming we know implementation details.
Assuming a blackbox approach (or inventive ways to break the code without breaking the test) this is not completely reliable.

Reply:
Maybe I'm missing something, but I think the whole point of `SETTING_ALLOCATION_MAX_RETRY` is to limit this value to 5, blocking further recovery attempts once it reaches the limit.

Reply:
Correct, as long as we can rely on it, this test is valid.